src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+    -- a 'SyncEvent' to 'm ()'.
+    }
+
+-- | A sync monad over @m@, that can throw 'SyncError's.
+newtype Sync c tx s m a =
+    Sync { runSync :: ExceptT SyncError (ReaderT (SyncContext c tx s m) m) a }
+    deriving ( Functor
+             , Applicative
+             , Monad
+             , MonadError SyncError
+             , MonadReader (SyncContext c tx s m)
+             )
+
+instance MonadTrans (Sync c tx s) where
+    lift = Sync . lift . lift
+
+instance MonadIO m => MonadIO (Sync c tx s m) where
+    liftIO = Sync . liftIO
+
+{------------------------------------------------------------------------------
+  Convenience functions over Sync
+------------------------------------------------------------------------------}
+
+withActivePeers
+    :: Monad m
+    => (ActivePeers c -> Sync c tx s m a)
+    -> Sync c tx s m a
+withActivePeers f = do
+    ctx <- ask
+    activePeers <- lift (scActivePeers ctx)
+    when (HS.null activePeers) $ throwError NoActivePeers
+    f activePeers
+
+-- | Like 'withActivePeers', but it passes to the callback only a (single)
+-- active peer. This might be useful in tests which it's not necessary to have
+-- all the active peers available.
+withActivePeer
+    :: Monad m
+    => (ActivePeer c -> Sync c tx s m a)
+    -> Sync c tx s m a
+withActivePeer f = do
+    ctx <- ask
+    activePeers <- lift (scActivePeers ctx)
+    case HS.toList activePeers of
+      []           -> throwError NoActivePeers
+      activePeer:_ -> f activePeer
+
+{------------------------------------------------------------------------------
+  Pure functions & constants
+------------------------------------------------------------------------------}
+
+height :: Block c tx s -> Height
+height = blockHeight . blockHeader
+
+-- | The nu parameter, from the spec.
+type Nu = Natural
+
+maxBlocksPerPeer :: Int
+maxBlocksPerPeer = 50
+
+-- | Returns 'True' if the node finished syncing all the blocks.
+-- This is the 'isDone' state transition function from the spec.
+isDone
+    :: ( Eq tx
+       , Eq s
+       , Eq (Hash c)
+       , Eq (PublicKey c)
+       )
+    => Nu
+    -> Block c tx s
+    -> Block c tx s
+    -> Bool
+isDone (fromIntegral -> nu) localTip remoteTip =
+    remoteTip == localTip || height remoteTip - height localTip < nu
+
+-- | Returns the range between two blocks.
+range
+    :: Block c tx s
+    -> Block c tx s
+    -> Maybe Range
+range localTip remoteTip
+  | outOfRange = Nothing
+  | otherwise  = Just (Range start end)
+  where
+    start      = succ (height localTip)
+    end        = height remoteTip
+    outOfRange = end < start
+
+{------------------------------------------------------------------------------
+  Monadic operations
+------------------------------------------------------------------------------}
+
+-- | The 'remoteTip' operation from the spec. It tries to fetch the tips
+-- concurrently, picking the tip returned by /most/ of the peers.
```

Comment:
If I understand, there's the possibility that each peer returns a different tip. If that's the case, what happens?

Reply:
We pick the longest, but I guess even "pick randomly" could have worked:

```
      comparingTips (b1,f1) (b2,f2) =
          case f2 `compare` f1 of
            EQ -> height b2 `compare` height b1
            x  -> x
```

The advantage of picking randomly is that you are (allegedly) less subject to attacks where a dishonest player would trick you into following his chain by forging a tip with a very high .. height (sorry for the repetition 😉 )